langchain/vectorstores/matching_engine.py

```diff
@@ -0,0 +1,433 @@
+"""Vertex Matching Engine implementation of the vector store."""
+from __future__ import annotations
+
+import json
+import logging
+import time
+import uuid
+from typing import TYPE_CHECKING, Any, Iterable, List, Optional, Type
+
+from langchain.docstore.document import Document
+from langchain.embeddings import TensorflowHubEmbeddings
+from langchain.embeddings.base import Embeddings
+from langchain.vectorstores.base import VectorStore
+
+if TYPE_CHECKING:
+    from google.cloud import storage
+    from google.cloud.aiplatform import MatchingEngineIndex, MatchingEngineIndexEndpoint
+    from google.oauth2.service_account import Credentials
+
+logger = logging.getLogger()
+
+
+class MatchingEngine(VectorStore):
+    """Vertex Matching Engine implementation of the vector store.
+
+    While the embeddings are stored in the Matching Engine, the embedded
+    documents will be stored in GCS.
+
+    An existing Index and corresponding Endpoint are preconditions for
+    using this module.
+
+    See usage in docs/modules/indexes/vectorstores/examples/matchingengine.ipynb
+
+    Note that this implementation is mostly meant for reading if you are
+    planning to do a real time implementation. While reading is a real time
+    operation, updating the index takes close to one hour."""
+
+    def __init__(
+        self,
+        project_id: str,
+        index: MatchingEngineIndex,
+        endpoint: MatchingEngineIndexEndpoint,
+        embedding: Embeddings,
+        gcs_client: storage.Client,
+        gcs_bucket_name: str,
+        credentials: Credentials = None,
+    ):
+        """Vertex Matching Engine implementation of the vector store.
+
+        While the embeddings are stored in the Matching Engine, the embedded
+        documents will be stored in GCS.
+
+        An existing Index and corresponding Endpoint are preconditions for
+        using this module.
+
+        See usage in
+        docs/modules/indexes/vectorstores/examples/matchingengine.ipynb.
+
+        Note that this implementation is mostly meant for reading if you are
+        planning to do a real time implementation. While reading is a real time
+        operation, updating the index takes close to one hour.
+
+        Attributes:
+            project_id: The GCS project id.
+            index: The created index class. See
+            ~:func:`MatchingEngine.from_components`.
+            endpoint: The created endpoint class. See
+            ~:func:`MatchingEngine.from_components`.
+            gcs_client: The Google Cloud Storage client.
```

Comment:
@hwchase17 maybe the attributes order should match the constructor arguments? Also I think `gcs_bucket_name` and `credentials` are missing.